### PR TITLE
drivers: sensor: max30208: Added driver and sample for MAX30208

### PR DESCRIPTION
**Pull request by trembel**
Added driver and sample for MAX30208 Maxim Integrated I2C temperature
sensor. Tested on nrf52dk.

Signed-off-by: Silvano Cortesi <silvano.cortesi@hotmail.com>

### CODE REVIEW

File: drivers/sensor/max30208/max30208.c [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+		return -EIO;
+	}
+#endif
+
+	return 0;
+}
+
+static struct max30208_config max30208_config = {
+	.i2c_label = DT_INST_BUS_LABEL(0),
+	.i2c_addr = DT_INST_REG_ADDR(0),
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:198]: better use i2c_dt_spec

File: drivers/sensor/max30208/max30208.c [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_FIFO_CFG1,
+			       config->fifo.fifo_a_full)) {
+		LOG_ERR("Failed to initialize FIFO config 1!");
+		return -EIO;
+	}
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_FIFO_CFG2,
+			       MAX30208_STC_TO_B(config->fifo.config2))) {
+		LOG_ERR("Failed to initialize FIFO config 2!");
+		return -EIO;
+	}
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:177]: error codes need to be propagated,

```c
ret = f();
if (ret < 0) {
    return ret;
}
```

File: drivers/sensor/max30208/max30208.c [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+	/* Reset the sensor */
+	if (i2c_reg_write_byte(data->i2c, config->i2c_addr, MAX30208_REG_SYS_CTRL,
+			       MAX30208_RESET_MASK)) {
+		LOG_ERR("Failed to reset sensor!");
+		return -EIO;
+	}
+	/* Wait for reset to be cleared */
+	k_sleep(K_MSEC(10));
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:158]: constant needs to be defined

File: drivers/sensor/max30208/max30208.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		LOG_ERR("Could not read sensor value");
+		return -EIO;
+	}
+
+	/* Concat two bytes to one sample */
+	data->raw = (int16_t)(local_buffer[0] << 8) | local_buffer[1];
+	return 0;
+}
+
+static int max30208_sample_fetch(const struct device *dev, enum sensor_channel chan)
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:66]: requested channel is not checked here (all or ambient temp)

File: drivers/sensor/max30208/max30208.c [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	uint8_t local_buffer[2] = { 0 };
+
+	/* Read data ready */
+	if (i2c_burst_read(data->i2c, config->i2c_addr, MAX30208_REG_FIFO_DATA, local_buffer, 2)) {
+		LOG_ERR("Could not read sensor value");
+		return -EIO;
+	}
+
+	/* Concat two bytes to one sample */
+	data->raw = (int16_t)(local_buffer[0] << 8) | local_buffer[1];
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:62]: please, use sys/byteorder.h API

File: drivers/sensor/max30208/Kconfig [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+config MAX30208_GPIO1_MODE
+	int "GPIO 1 mode"
+	range 0 3
+	default 2
+	help
+	    Set the mode for GPIO1.
+	    0 = High impedance input
+	    1 = Output (open-drain)
+	    2 = Input with 1MOhm pulldown
+	    3 = Convert Temperature input (active low)
```
Comment (gmarull) [inline drivers/sensor/max30208/Kconfig:62]: these options can likely be moved to devicetree, so that they can be configured per-instance.

File: drivers/sensor/max30208/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+#
+# Copyright (c) 2021, Silvano Cortesi
+# Copyright (c) 2021, ETH Zürich
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+zephyr_library()
+
+zephyr_library_sources(max30208.c)
+zephyr_library_sources_ifdef(CONFIG_MAX30208_TRIGGER    max30208_trigger.c)
```
Comment (gmarull) [inline drivers/sensor/max30208/CMakeLists.txt:11]:
```suggestion
zephyr_library_sources_ifdef(CONFIG_MAX30208_TRIGGER max30208_trigger.c)
```

File: drivers/sensor/max30208/max30208.c [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+#ifdef CONFIG_MAX30208_TRIGGER
+	.int_pin = DT_INST_GPIO_PIN(0, int_gpios),
+	.int_flags = DT_INST_GPIO_FLAGS(0, int_gpios),
+	.int_name = DT_INST_GPIO_LABEL(0, int_gpios),
+#endif
+};
+
+static struct max30208_data max30208_data;
+
+DEVICE_DT_INST_DEFINE(0, max30208_init, NULL, &max30208_data, &max30208_config, POST_KERNEL,
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:233]: please, make driver multi-instance (see most other sensors)

File: drivers/sensor/max30208/max30208.h [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+	MAX30208_GPIO_INPUT = 0,
+	MAX30208_GPIO_OUTPUT = 1,
+	MAX30208_GPIO_INPUT_PULLDOWN = 2,
+	MAX30208_GPIO_CONV_TEMP = 3,
+	MAX30208_GPIO_INTB = 3,
+};
+
+/* register structs */
+#ifdef CONFIG_BIG_ENDIAN
+struct MAX30208_status {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.h:88]: naming convention:
```suggestion
struct max30208_status {
```

File: drivers/sensor/max30208/max30208_trigger.c [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+	gpio_init_callback(&data->gpio_cb, max30208_gpio_callback, BIT(config->int_pin));
+
+	int rc = gpio_pin_configure(data->gpio, config->int_pin, GPIO_INPUT | config->int_flags);
+
+	if (rc == 0) {
+		gpio_add_callback(data->gpio, &data->gpio_cb);
+	}
+
+	if (rc != 0) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:222]: this can be moved above, so `rc == 0` branch is not required

File: drivers/sensor/max30208/max30208_trigger.c [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+		LOG_DBG("Failed to get pointer to %s device!", config->int_name);
+		return -EINVAL;
+	}
+
+	gpio_init_callback(&data->gpio_cb, max30208_gpio_callback, BIT(config->int_pin));
+
+	int rc = gpio_pin_configure(data->gpio, config->int_pin, GPIO_INPUT | config->int_flags);
+
+	if (rc == 0) {
+		gpio_add_callback(data->gpio, &data->gpio_cb);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:219]: error check missing here

File: drivers/sensor/max30208/max30208_trigger.c [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+int max30208_init_interrupt(const struct device *dev)
+{
+	struct max30208_data *data = dev->data;
+	const struct max30208_config *config = dev->config;
+
+	data->status = 0;
+
+	data->gpio = device_get_binding(config->int_name);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:208]: please, move to `gpio_dt_spec`

File: samples/sensor/max30208/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
Comment (gmarull) [inline samples/sensor/max30208/CMakeLists.txt:3]: sample should go into its own separate commit

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for this addition, left a few comments

File: drivers/sensor/max30208/Kconfig (hunk 6 above)
Comment (trembel) [reply drivers/sensor/max30208/Kconfig:62]: can they be moved to dev tree but still have a Kconfig being dependend on them (interrupt requires GPIO0 mode being 3)?

File: drivers/sensor/max30208/max30208.h (hunk 9 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.h:88]: what for enums? also lower case for its name (and then inside upper case)?

File: drivers/sensor/max30208/max30208.c (hunk 1 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.c:198]: I see. Dev began before the addition of `i2c_dt_spec`. How can I rebase my branch without losing my changes to a newer version?

File: drivers/sensor/max30208/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# Makefile - MAX30208 digital temperature sensor
+#
```
Comment (gmarull) [inline drivers/sensor/max30208/CMakeLists.txt:2]: nit: not a Makefile :-)

File: drivers/sensor/max30208/max30208.c [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+		return ret;
+	}
+
+	return 0;
+}
+
+static int max30208_channel_get(const struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:35]: `SENSOR_CHAN_ALL` should also be supported, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

File: drivers/sensor/max30208/max30208.c [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+		return -ENOTSUP;
+	}
+	struct max30208_data *data = dev->data;
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL] = { 0 };
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:39]: better move declarations/init to the top.

File: drivers/sensor/max30208/max30208.c [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+
+#ifndef CONFIG_MAX30208_TRIGGER
+static int max30208_data_ready(const struct device *dev, uint8_t *ready)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_status status;
+
+	/* Read data ready */
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_INT_STS, MAX30208_STC_TO_P(status));
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:70]: doesn't seem to follow styleguide:
```
int ret = f(a, very, long,
            list);
```

File: drivers/sensor/max30208/max30208.c [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+
+	if (ret < 0) {
+		LOG_ERR("Timeout while waiting on MAX30208 data ready\n");
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:73]: `ret < 0` from `i2c_reg_read_byte_dt` doesn't imply a timeout, afaik.

File: drivers/sensor/max30208/max30208.c [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0;
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_FIFO_DATA_CTR, &fifo_data_counter);
+
+	if (ret < 0) {
+		LOG_ERR("Timeout while waiting on FIFO data counter\n");
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:93]: ditto

File: drivers/sensor/max30208/max30208.h [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+#define MAX30208_ONE_DEGREE 0xC8
+#define MAX30208_LSB_e6 5000
+
+/* enums and definitions */
+#define MAX30208_RESET_MASK 0x01
+#define MAX30208_CONVERT_T_MASK 0xFF
+#define MAX30208_INT_A_FULL_MASK (1 << 7)
+#define MAX30208_INT_TEMP_LO_MASK (1 << 2)
+#define MAX30208_INT_TEMP_HI_MASK (1 << 1)
+#define MAX30208_INT_TEMP_RDY_MASK (1 << 0)
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.h:73]: BIT() macro should be used

File: drivers/sensor/max30208/max30208.h [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+#define MAX30208_GPIO_MODE_MASK(gpioX, mode) (mode << (gpioX * 6))
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.h:74]: `6U`

File: drivers/sensor/max30208/max30208.c [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+	val->val2 = (temp_val % MAX30208_ONE_DEGREE) * MAX30208_LSB_e6;
+
+	return 0;
+}
+
+#ifndef CONFIG_MAX30208_TRIGGER
+static int max30208_data_ready(const struct device *dev, uint8_t *ready)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_status status;
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:66]: why using a struct to read a single byte?

File: drivers/sensor/max30208/max30208.c [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+	return 0;
+}
+#endif
+
+int max30208_readout_batch(const struct device *dev)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0;
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:88]: `0U`

File: drivers/sensor/max30208/max30208.c [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+	for (uint8_t i = 0; i < fifo_data_counter; i++) {
+		if (ring_buf_space_get(&data->raw_buffer) < MAX30208_BYTES_PER_VAL) {
+			struct sensor_value lost_val;
+
+			ret = max30208_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &lost_val);
+			if (ret < 0) {
+				return ret;
+			}
+			LOG_INF("Buffer size to small. Value %d.%d is lost", lost_val.val1,
+				lost_val.val2);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:118]: is it worth reading just to log? what if logging is disabled? Isn't it sufficient to warn about buffer being too small?

File: drivers/sensor/max30208/max30208.c [hunk 25]
```diff
@@ -0,0 +1,346 @@
...
+#define MAX30208_A_FULL_TYPE .a_full_type = CONFIG_MAX30208_FIFO_A_FULL_TYPE,
+#else
+#define MAX30208_A_FULL_TYPE .a_full_type = 0,
+#endif
+
+#ifdef CONFIG_MAX30208_FIFO_STAT_CLR
+#define MAX30208_FIFO_STAT_CLR .fifo_stat_clr = CONFIG_MAX30208_FIFO_STAT_CLR,
+#else
+#define MAX30208_FIFO_STAT_CLR .fifo_stat_clr = 0,
+#endif
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:309]: Kconfig options can be used directly without storing them in config. Why not make these DT properties?, so they can be set per-instance?

File: drivers/sensor/max30208/max30208.c [hunk 26]
```diff
@@ -0,0 +1,346 @@
...
+	    .gpio_setup = {	\
+		.gpio0_mode = DT_INST_PROP(n, gpio0_mode),	\
+		.gpio1_mode = DT_INST_PROP(n, gpio1_mode),	\
+	    },	\
+	MAX30208_GPIO_TRIGGER(n)	\
+	};	\
+	\
+	static struct max30208_data max30208_data_##n = {	\
+		.raw_buffer = {	\
+			.size = CONFIG_MAX30208_RINGBUFFER_SIZE * MAX30208_BYTES_PER_VAL,	\
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:338]: `ARRAY_SIZE`?

File: drivers/sensor/max30208/max30208_trigger.c [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+	return 0;
+}
+
+static void setup_int(const struct device *dev, bool enable)
+{
+	const struct max30208_config *config = dev->config;
+
+	gpio_flags_t flags = enable ? GPIO_INT_EDGE_TO_ACTIVE : GPIO_INT_DISABLE;
+
+	gpio_pin_interrupt_configure_dt(&config->gpio_int, flags);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:69]: this returns error codes, should be checked.

File: drivers/sensor/max30208/max30208_trigger.c [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+
+	if (!device_is_ready(config->gpio_int.port)) {
+		LOG_DBG("GPIO device pointer is not ready!");
+		return -ENODEV;
+	}
+
+	int ret = gpio_pin_configure_dt(&config->gpio_int, GPIO_INPUT);
+
+	if (ret < 0) {
+		LOG_DBG("Failed to set gpio as input!");
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:220]: `LOG_ERR`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/max30208/max30208.c (hunk 25 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.c:309]: would you make all of my Kconfig options to DT properties?

File: drivers/sensor/max30208/max30208.c (hunk 24 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.c:118]: since one item must be poped on a full ringbuffer, I could just pop and not store it, but I don't see a lot of advantage here..

File: drivers/sensor/max30208/max30208.c (hunk 15 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.c:35]: but this is channel get, shouldn't it be specified here?

File: drivers/sensor/max30208/max30208.c (hunk 22 above)
Comment (trembel) [reply drivers/sensor/max30208/max30208.c:66]: Its a bitfield struct. As it implies the same runtime cost as using a raw byte, I think its better here as specific fields can be accessed by a meaningful name.

**Comment by trembel** [comment]
Open Questions to me:
> `SENSOR_CHAN_ALL` should also be supported, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

As written, should this also be for `channel_get`? IMO not.

> is it worth reading just to log? what if logging is disabled? Isn't it sufficient to warn about buffer being too small?

since one item must be poped on a full ringbuffer, I could just pop and not store it, but I don't see a lot of advantage here..

> why using a struct to read a single byte?

Its a bitfield struct. As it implies the same runtime cost as using a raw byte, I think its better here as specific fields can be accessed by a meaningful name.

**Comment by carlescufi** [comment]
@gmarull can you please take another look?

File: drivers/sensor/max30208/max30208.c [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		return -ENOTSUP;
+	}
+
+	/* If new value in ring buffer available, get it */
+	if (ring_buf_size_get(&data->raw_buffer) < MAX30208_BYTES_PER_VAL) {
+		return -ENODATA;
+	}
+
+	int ret = ring_buf_get(&data->raw_buffer, local_buffer, MAX30208_BYTES_PER_VAL);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:47]: this function returns `uint32_t`, so error handling here is incorrect. ref https://docs.zephyrproject.org/latest/reference/data_structures/ring_buffers.html#c.ring_buf_get

File: drivers/sensor/max30208/max30208.c [hunk 30]
```diff
@@ -0,0 +1,329 @@
...
+#endif
+
+int max30208_readout_batch(const struct device *dev)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0U;
+	int ret =
+		i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_FIFO_DATA_CTR, &fifo_data_counter);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:91]: formatting can be improved here:
```
	int ret = i2c_reg_read_byte_dt(&config->bus,
				       MAX30208_REG_FIFO_DATA_CTR,
				       &fifo_data_counter);
```

File: drivers/sensor/max30208/max30208.c [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+
+	if (ret < 0) {
+		LOG_ERR("Could not read from MAX30208");
+		return ret;
+	}
+
+	/* Read data ready */
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL * MAX30208_FIFO_SIZE];
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:99]: define at the top of the function (per convention)

File: drivers/sensor/max30208/max30208.c [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+	k_sleep(K_MSEC(MAX30208_TMP_MEAS_TIME));
+
+	/* Poll until conversion finishes */
+	for (uint8_t i = 0, ready = 0; ready == 0; i++) {
+		ret = max30208_data_ready(dev, &ready);
+
+		if (ret < 0) {
+			return ret;
+		}
+		if (i == MAX30208_POLL_TRIES) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:194]: you'll be running `MAX30208_POLL_TRIES + 1` this way

File: drivers/sensor/max30208/max30208.c [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+	}
+
+	/* Check the part id to make sure this is MAX30208 */
+	int ret = i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_PART_ID, &part_id);
+
+	if (ret < 0) {
+		LOG_ERR("Could not get Part ID");
+		return ret;
+	}
+	if (part_id != MAX30208_PART_ID) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:240]: nit: blank line here

File: drivers/sensor/max30208/max30208.c [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+		MAX30208_GPIO_TRIGGER(n)	\
+	};	\
+	\
+	static struct max30208_data max30208_data_##n = {	\
+		.raw_buffer = {	\
+			.size = MAX30208_ARRAY_SIZE,	\
+			.buf = { .buf8 = max30208_data_##n._ring_buffer_data_raw_buffer }	\
+		}	\
+	};	\
+	\
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:325]: nit: push all backslashes to the same column

File: dts/bindings/sensor/maxim,max30208.yaml [hunk 35]
```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2021, Silvano Cortesi
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/maxim,max30208.yaml:1]: This binding needs to justify its use of defaults; see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#general-rules for details.

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
+1 for the binding. I didn't review the rest.

**Comment by trembel** [comment]
what is blocking this one?

File: drivers/sensor/max30208/max30208.c [hunk 36]
```diff
@@ -0,0 +1,323 @@
...
+	const struct max30208_config *config = dev->config;
+
+	/* Trigger one measurement */
+	int ret = i2c_reg_write_byte_dt(&config->bus, MAX30208_REG_TEMP_SETUP,
+					MAX30208_CONVERT_T_MASK);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:28]: nit: just return here

File: drivers/sensor/max30208/max30208.c [hunk 37]
```diff
@@ -0,0 +1,323 @@
...
+
+	/* Concat two bytes to a int16 value */
+	int16_t temp_val = (int16_t)sys_get_be16(local_buffer);
+
+	/* Convert raw value into temperature */
+	val->val1 = temp_val / MAX30208_ONE_DEGREE;
+	val->val2 = (temp_val % MAX30208_ONE_DEGREE) * MAX30208_LSB_e6;
+
+	return 0;
+}
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:60]: this will succeed even if there is no data

File: drivers/sensor/max30208/max30208.c [hunk 38]
```diff
@@ -0,0 +1,323 @@
...
+
+#ifndef CONFIG_MAX30208_TRIGGER
+static int max30208_data_ready(const struct device *dev, uint8_t *ready)
+{
+	const struct max30208_config *config = dev->config;
+	struct max30208_status status;
+
+	/* Read data ready */
+	int ret = i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_INT_STS,
+					MAX30208_STC_TO_P(status));
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:70]: nit: align second line to character after `(`

File: drivers/sensor/max30208/max30208.c [hunk 39]
```diff
@@ -0,0 +1,323 @@
...
+	const struct max30208_config *config = dev->config;
+	struct max30208_data *data = dev->data;
+
+	uint8_t fifo_data_counter = 0U;
+	uint8_t local_buffer[MAX30208_BYTES_PER_VAL * MAX30208_FIFO_SIZE];
+
+	/* Read data ready */
+	int ret = i2c_reg_read_byte_dt(&config->bus,
+					MAX30208_REG_FIFO_DATA_CTR,
+					&fifo_data_counter);
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:93]: nit: align to `(`

File: drivers/sensor/max30208/max30208.c [hunk 40]
```diff
@@ -0,0 +1,323 @@
...
+	/* Wait for I2C bus */
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("I2C dev %s not ready", config->bus.bus->name);
+		return -ENODEV;
+	}
+
+	/* Check the part id to make sure this is MAX30208 */
+	int ret = i2c_reg_read_byte_dt(&config->bus, MAX30208_REG_PART_ID, &part_id);
+
+	if (ret < 0) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:229]: this, an others: blank line not needed (please be consistent)

File: drivers/sensor/max30208/max30208.c [hunk 41]
```diff
@@ -0,0 +1,323 @@
...
+				.fifo_stat_clr = DT_INST_PROP(n, fifo_stat_clr),		\
+				.flush_fifo = 0,						\
+			},									\
+		},										\
+	    .gpio_setup = {									\
+		.gpio0_mode = DT_INST_PROP(n, gpio0_mode),					\
+		.gpio1_mode = DT_INST_PROP(n, gpio1_mode),					\
+	    },											\
+		MAX30208_GPIO_TRIGGER(n)							\
+	};											\
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208.c:311]: nit: mixes spaces/tabs

**Review by gmarull — COMMENTED** [review]
Some minor comments, LGTM otherwise

File: samples/sensor/max30208/sample.yaml [hunk 42]
```diff
@@ -0,0 +1,11 @@
+sample:
+  description: TBD
+  name: TBD
```
Comment (gmarull) [inline samples/sensor/max30208/sample.yaml:3]: needs update

File: samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay [hunk 43]
```diff
@@ -0,0 +1,17 @@
...
+    max30208@50 {
+        compatible = "maxim,max30208";
+        reg = <0x50>;
+        label = "MAX30208";
+        int-gpios = <&gpio0 22 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+	fifo-a-full = <2>;
+/*      gpio0-mode = <3>; */
+	status = "okay";
+    };
+};
```
Comment (gmarull) [inline samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay:17]: mixes tabs and spaces, why commenting `gpio0-mode`?

File: samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay [hunk 44]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2019, Linaro Ltd.
```
Comment (gmarull) [inline samples/sensor/max30208/boards/nrf52dk_nrf52832.overlay:2]: `Linaro`?

File: samples/sensor/max30208/sample.yaml [hunk 45]
```diff
@@ -0,0 +1,11 @@
...
+tests:
+  sample.sensor.max30208:
+    harness: sensor
+    tags: sensors
+    platform_allow: nrf52dk_nrf52832
```
Comment (gmarull) [inline samples/sensor/max30208/sample.yaml:8]: This sample seems to be generic enough, so it can run on more boards. Better use `filter: dt_compat_enabled("maxim,max30208")`. See for example `samples/sensor/lps22hh/sample.yaml`.

File: drivers/sensor/max30208/max30208_trigger.c [hunk 46]
```diff
@@ -0,0 +1,263 @@
...
+
+	/* Write the alarm configuration register */
+	int ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[1], (uint8_t)((temp >> 8) & 0x00FF));
+
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize Alarm MSB!");
+		return ret;
+	}
+	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[0], (uint8_t)(temp & 0x00FF));
+	if (ret < 0) {
```
Comment (gmarull) [inline drivers/sensor/max30208/max30208_trigger.c:55]: there seems to be inconsistencies still, please revisit

```c
	int ret;

	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[1], (uint8_t)((temp >> 8) & 0x00FF));
	if (ret < 0) {
		LOG_ERR("Failed to initialize Alarm MSB!");
		return ret;
	}

	ret = i2c_reg_write_byte_dt(&config->bus, reg_addr[0], (uint8_t)(temp & 0x00FF));
	if (ret < 0) {
```

**Review by gmarull — CHANGES_REQUESTED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
for the DT portions

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Sorry for the late review. Overall this looks very good. It needs a rebase and to add the new sensor to `tests/drivers/build_all/sensor`
